### ISSUE TITLE
nn.Embedding padding idx doesn't check for negative value

### ISSUE THREAD

**Issue by cedias** [pytorch/pytorch#3233]
nn.Embedding doesn't check for negative padding index value which might be able to lead to subtle bugs if someone wants to use it with inverted indexing (even though it's not specified that you can).

```
def forward(self, input):
        padding_idx = self.padding_idx
        if padding_idx is None:
            padding_idx = -1
        return self._backend.Embedding.apply(
            input, self.weight,
            padding_idx, self.max_norm, self.norm_type,
            self.scale_grad_by_freq, self.sparse
        )

```

To be clear: defining an embedding layer `emb = nn.Embedding(5,10,padding_idx=-1)` is possible and raises no error. Moreover the last weights are initialized to 0 as you would expect if negative indexing on embedding was working. This can be confusing. 

```
In [44]: emb(Variable(torch.LongTensor([4])))
Out[44]: 
Variable containing:
    0     0     0     0     0     0     0     0     0     0
[torch.FloatTensor of size 1x10]
```

I believe this can be fixed in two ways:

- asserting `padding_idx > 0 or None`
- if `padding_idx < 0` then setting `padding_idx = num_embedding + padding_idx`
